Make `layer_predict` forward stored dots_list to `predict()`

brookslogan opened the pull request:
### Checklist

Please:

- [x] Make sure this PR is against "dev", not "main".
- [x] Request a review from one of the current epipredict main reviewers:
      dajmcdon.
- [x] Make sure to bump the version number in `DESCRIPTION` and `NEWS.md`.
      Always increment the patch version number (the third number), unless you are
      making a release PR from dev to main, in which case increment the minor
      version number (the second number).
- [x] Describe changes made in NEWS.md, making sure breaking changes
      (backwards-incompatible changes to the documented interface) are noted.
      Collect the changes under the next release number (e.g. if you are on
      0.7.2, then write your changes under the 0.8 heading).
- [x] Consider pinning the `epiprocess` version in the `DESCRIPTION` file if
  - You anticipate breaking changes in `epiprocess` soon
  - You want to co-develop features in `epipredict` and `epiprocess`

### Change explanations for reviewer

This uses `rlang::inject()` to splat in the stored dots list.  Some validation and tests have also been added, plus following @dajmcdon's suggested correction to an `@inheritParams` directive.   

### Magic GitHub syntax to mark associated Issue(s) as resolved when this is merged into the default branch

- Resolves #46

dajmcdon commented on tests/testthat/test-layer_predict.R:
```diff
@@ -31,3 +31,47 @@ test_that("prediction with interval works", {
+    mutate(sandwiched = .pred_lower <= .pred & .pred <= .pred_upper) %>%
+    `[[`("sandwiched") %>%
+    all())
+  # There are many possible other valid configurations that aren't tested here.
```
One potential further issue (possibly out of scope) is the one RyanN uncovered: if you manage to get your workflow to fit multiple response variables simultaneously, the `type = "pred_int"` chokes. 

``` r
library(workflows)
library(recipes)

tib <- tibble(y1 = rnorm(100), y2 = rnorm(100), x = rnorm(100), z = rnorm(100))
rec <- recipe(y1 + y2 ~ x + z, data = tib) %>%
  step_normalize(all_numeric_predictors())
wf <- workflow(rec, parsnip::linear_reg()) %>% generics::fit(tib)
predict(wf, tib[1:5, ])
#> # A tibble: 5 × 2
#>   .pred_y1 .pred_y2
#>      <dbl>    <dbl>
#> 1    0.285 -0.0736 
#> 2   -0.102 -0.0145 
#> 3   -0.206 -0.00517
#> 4   -0.180 -0.00994
#> 5   -0.256  0.0434
predict(wf, tib[1:5, ], type = "pred_int")
#> Error in `dplyr::select()`:
#> ! Can't subset columns that don't exist.
#> ✖ Column `fit` doesn't exist.
```

brookslogan replied:
Maybe related, maybe not: I just remembered that you can get `predict(ewf, dat, type = "pred_int")` to work in some cases... but I haven't made `predict(ewf, dat, type = "pred_int", level = 0.8)` work.  Do we want to make the former an error, or the latter work?

brookslogan replied:
That seems like an upstream bug or bugs that we also trip over....
- parsnip either generates an interval only for y2, or generates this error, depending on whether you write `y1 + y2 ~` or `cbind(y1, y2) ~`
- recipes tries to prevent you from writing `cbind(y1, y2) ~`
- workflows transforms a recipe formula from `y1 + y2 ~` to `cbind(y1, y2) ~`

We could attempt to do better here by trying to detect & either abort or patch over these bad cases... but should we just be filing/fixing upstream?

brookslogan commented on tests/testthat/test-layer_predict.R:
```diff
@@ -31,3 +31,47 @@ test_that("prediction with interval works", {
+
+  # ^ (currently with a truly awful error message, due to an extra comma in parsnip::check_pred_type_dots)
+  #
+  # Unfortunately, we outright ignore attempts to pass args via `predict.epi_workflow`:
+  f_predict <- frosting() %>% layer_predict()
+  wf_predict <- wf %>% add_frosting(f_predict)
```
Do we want `...` from `predict.epi_workflow()` to be forwarded only to `slather.layer_predict()` or to all the other layer `slather` methods?  Would other layers ever need to know about these args?

(If we do end up forwarding to all `slather.layer_*`, we should make sure the only args allowed in `slather.layer_*` are documented (correctly) in `predict.epi_workflow()` to prevent name collisions from inducing crazy behavior.)

.... Just from thinking about this, I think there's probably a bunch of errors in the current implementation, and there's going to be a bunch of extra logic that might be required.  E.g., for `layer_population_scaling` to work correctly, it needs to know what the prediction output columns are.  But that changes based on the prediction type!  Do we have roles of columns tracked in the prediction output that might allow this to be automated?  I assume not, right?

brookslogan replied:
Do we want to continue down this sort of path & fix up issues, or instead disallow `type` etc. and make them use layers instead?

dajmcdon replied:
If we forward, I think only to `layer_predict()`. And perhaps only in the case that that's the only layer. Those arguments could potentially result in arbitrary structures in the predictions data frame, so I think it would be near impossible to anticipate them all. Perhaps additional layers could be written specifically to handle some of these options in the future.

brookslogan replied:
Okay, makes sense.  For this PR, I'll probably hack on the "only in the case that that's the only layer" part in a minimal fashion.

brainstorm: if we add processing to other layers for these, we probably want a bit of refactoring:
- Add an extra required argument to `slather()` to hold the dots list, instead of passing via dots.
- Pass dots list to every layer.
- Validate layer support by either:
  - Checking for an empty dots list in each unsupported `slather()` --- initially, we'd add such a check at the beginning of every `slather.layer_*()` other than `slather.layer_predict()`.
  - Have an S3 generic (e.g., "`predict_dots_supported(layer)`") that outputs whether/which predict dots arguments are supported, and use that in `predict.epi_workflow()`.  This is more complex but could list all the unsupported layers at once and may help us/devs not forget to check for emptiness.
  - Have a fixed list of layers with support that we check against in `predict.epi_workflow()`.  Not extensible.

brookslogan replied:
Should this settings-forwarding also be supported by `forecast`?  Or do we want to push people toward more-completely-supported features there?

dajmcdon replied:
Let's say "completely supported features for now"

brookslogan replied:
todo: looks like code comment above shouldn't apply now; revise and check if test is also wrong now

brookslogan commented:
@dajmcdon I addressed some comments but still have a few pending questions up above.  The new changes might be worth a look... they don't seem too elegant as (i) they involve passing around type, opts, ... in many places but not everywhere, and (ii) because it's actually forwarding stuff in previously-unused ...; it was a quick and NSE-compatible way to do things, but I'm not sure we want to sacrifice parts of the dots "namespace" in so many places.

dajmcdon reviewed (approved):
Hi @brookslogan this all looks good to me. Feel free to merge.

brookslogan commented:
@dajmcdon tests were failing due to an [epiprocess update](https://github.com/cmu-delphi/epiprocess/pull/472); I've slightly changed bake behavior in a way I think is a rare bugfix.  Since bake-ing's pretty key, I'll leave it to you to merge.
